File: LinkedList/236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private Map<TreeNode, TreeNode> map = new HashMap<>();

    public  void getParentMap(TreeNode root) {
        if(root==null) return;
        if(root.left!=null) {
            map.put(root.left, root);
            getParentMap(root.left);
        }
        if(root.right!=null) {
            map.put(root.right, root);
            getParentMap(root.right);
        }


    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        
        if (root == null) return null;
        map.put(root, null);
        getParentMap(root);

        TreeNode curr = p;

        Set<TreeNode> pset = new HashSet<>();
        while(curr!=null) {
            pset.add(curr);
            curr = map.get(curr);
        }

        curr = q;

        while(curr!=null) {
            if(pset.contains(curr)) {
                return curr;
            }
            curr = map.get(curr);
        }
        return null;
    }
}

File: Array/236-lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        parent_map = {root: None}
        def getParentMap(node, parent):
            if node:
                if node.left:
                    parent_map[node.left] = node
                    getParentMap(node.left, node)
                if node.right:
                    parent_map[node.right] = node
                    getParentMap(node.right, node)
        
        getParentMap(root, None)

        path_nodes_p = set()

        curr = p
        while curr:
            path_nodes_p.add(curr)
            curr = parent_map[curr]
        
        curr = q
        while curr and curr not in path_nodes_p:
            curr = parent_map[curr]
        
        return curr


